File: SwrveSDK/SDK/Messaging/SwrveInAppStoryUIButton.m
#import "SwrveInAppStoryUIButton.h"
#import "SwrveSDKUtils.h"

#if __has_include(<SwrveSDKCommon/SwrveUtils.h>)
#import <SwrveSDKCommon/SwrveUtils.h>
#else
#import "SwrveUtils.h"
#endif

@implementation SwrveInAppStoryUIButton

@synthesize storyDismissButton;

- (id)   initWithButton:(SwrveStoryDismissButton *)dismissButton
           dismissImage:(UIImage *)dismissImage
dismissImageHighlighted:(UIImage *)dismissImageHighlighted {
    self = [super initWithFrame:CGRectZero]; // Frame will be set by the layout engine
    if (self) {
        self.storyDismissButton = dismissButton;

        [self setImage:dismissImage forState:UIControlStateNormal];
        if (dismissImageHighlighted) {
            [self setImage:dismissImageHighlighted forState:UIControlStateHighlighted];
        }
        self.tintColor = [SwrveUtils processHexColorValue:dismissButton.color];
        self.translatesAutoresizingMaskIntoConstraints = NO;
        self.imageView.contentMode = UIViewContentModeScaleToFill;
        self.contentHorizontalAlignment = UIControlContentHorizontalAlignmentFill;
        self.contentVerticalAlignment = UIControlContentVerticalAlignmentFill;
        self.adjustsImageWhenHighlighted = NO;
        self.isAccessibilityElement = true;
        self.accessibilityLabel = dismissButton.accessibilityText;
        self.accessibilityHint = @"Button";
    }
    return self;
}

- (void)setHighlighted:(BOOL)highlighted {
    [super setHighlighted:highlighted];
    if (highlighted) {
        self.tintColor = [SwrveUtils processHexColorValue:self.storyDismissButton.pressedColor];
    } else {
        self.tintColor = [SwrveUtils processHexColorValue:self.storyDismissButton.color];
    }
}

@end

File: SwrveSDKSwift/SwrveInAppStoryUIButton.swift
import UIKit

@objc public class SwrveInAppStoryUIButton: UIButton {

    @objc public var storyDismissButton: SwrveStoryDismissButton

    @objc(initWithButton:dismissImage:dismissImageHighlighted:)
    public init(
        button: SwrveStoryDismissButton,
        dismissImage: UIImage,
        dismissImageHighlighted: UIImage?
    ) {

        self.storyDismissButton = button
        super.init(frame: .zero)

        self.setImage(dismissImage, for: .normal)
        if let highlightedImage = dismissImageHighlighted {
            self.setImage(highlightedImage, for: .highlighted)
        }

        self.tintColor = SwrveUtils.processHexColorValue(button.color)
        self.translatesAutoresizingMaskIntoConstraints = false
        self.imageView?.contentMode = .scaleToFill
        self.contentHorizontalAlignment = .fill
        self.contentVerticalAlignment = .fill
        self.adjustsImageWhenHighlighted = false

        // Accessibility setup
        self.isAccessibilityElement = true
        self.accessibilityLabel = button.accessibilityText
        self.accessibilityHint = "Button"
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    @objc public override var isHighlighted: Bool {
        didSet {
            if isHighlighted {
                if let pressedColor = storyDismissButton.pressedColor {
                    self.tintColor = SwrveUtils.processHexColorValue(pressedColor)
                }
            } else if let normalColor = storyDismissButton.color {
                self.tintColor = SwrveUtils.processHexColorValue(normalColor)
            }
        }
    }
}
